File: Programmers/Lv3/change_word.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int ans = 100;

void dfs(string word, string target, vector<string> words,vector<bool> &visited,int answer){
    for(int i=0 ; i < words.size(); i++){
        int cnt = 0;
        /*사용하지 않은 단어 중 다른 글자수 check*/
        for(int j=0; j < word.length(); j++){
            if(!visited[i] && word[j] != words[i][j]) cnt++;
        }
        /*1글자만 다르다면*/
        if(cnt == 1){
            if(target == words[i] && ans > answer ){
                ans = answer + 1;
                return;
            }
            visited[i] = true;
            dfs(words[i],target,words,visited,answer+1);    
            visited[i] = false;
        }
    }
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    vector<bool> visited(words.size(),false);
    
    dfs(begin,target,words,visited,answer);
    
    return  (ans == 100) ? 0 : ans;
}

File: java/src/programmers/lv3/ChangeWord.java
package programmers.lv3;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

public class ChangeWord {
    boolean[] visited;
    int answer = 50;

    public int solution(String begin, String target, String[] words) {
        visited = new boolean[words.length];
        if (!Arrays.asList(words).contains(target)) return 0;

        for (int i = 0; i < words.length; i++) {
            if (checkWord(begin, words[i])) {
                visited[i] = true;
                dfs(words[i], target, 1, words);
                visited[i] = false;
            }
        }
        return answer;
    }

    private void dfs(String word, String target, int cnt, String[] words) {
        if (word.equals(target)) {
            answer = Math.min(cnt, answer);
            return;
        }

        for (int i = 0; i < words.length; i++) {
            if (!visited[i] && checkWord(word, words[i])) {
                visited[i] = true;
                dfs(words[i], target, cnt + 1, words);
                visited[i] = false;
            }
        }
    }

    private boolean checkWord(String str, String temp) {
        int cnt = 0;
        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) != temp.charAt(i)) {
                cnt++;
            }
            if (cnt > 1) return false;
        }
        return true;
    }
    @Test
    void test(){
        Assertions.assertEquals(4, solution("hit","cog",new String[]{"hot", "dot", "dog", "lot", "log", "cog"}));
    }
}
